packages/langium/src/grammar/type-system/inferred-types.ts

```diff
@@ -184,14 +184,23 @@ function addAction(state: TypeCollectorState, action: Action): void {
 function addAssignment(state: TypeCollectorState, assignment: Assignment): void {
     const typeItems: TypeCollection = { types: [], reference: false };
     findTypes(assignment.terminal, typeItems);
-    state.currentType?.properties.push({
```

Comment:
I don't see why this change should be necessary. It works correctly without the change.

Reply:
It improves the error message.

Before this change inferred types group properties types, but declared ones not. For example, here type of `x` in the declared type is an array of 3 `PropertyType`, but type of `x` in the inferred type is the only 1 `PropertyType`. It doesn't allow to match `PropertyType`s correctly and find out that the problem is only in type `number` (it's an array for the declared type but a primitive type for the inferred one).
```
interface X {
    x: string | AbstractDefinition | number[]
}

RuleX returns X:
    x=(ID | NUMBER | AbstractDefinition);
```
We got such error message before this update:
```
... 'AbstractDefinition | number | string' is not expected.
```
And get such one after the update:
```
... 'number' has to be an array.
```